examples/pytorch/graphsage/train_sampling_multi_gpu.py

```diff
@@ -73,32 +91,48 @@ def run(proc_id, n_gpus, args, devices, data):
         train_nfeat = val_nfeat = test_nfeat = g.ndata.pop('features')
         train_labels = val_labels = test_labels = g.ndata.pop('labels')
 
-    if not args.data_cpu:
-        train_nfeat = train_nfeat.to(dev_id)
-        train_labels = train_labels.to(dev_id)
-
     in_feats = train_nfeat.shape[1]
 
     train_mask = train_g.ndata['train_mask']
     val_mask = val_g.ndata['val_mask']
-    test_mask = ~(test_g.ndata['train_mask'] | test_g.ndata['val_mask'])
+    test_mask = test_g.ndata['test_mask']
     train_nid = train_mask.nonzero().squeeze()
     val_nid = val_mask.nonzero().squeeze()
     test_nid = test_mask.nonzero().squeeze()
 
     # Create PyTorch DataLoader for constructing blocks
     sampler = dgl.dataloading.MultiLayerNeighborSampler(
         [int(fanout) for fanout in args.fan_out.split(',')])
-    dataloader = dgl.dataloading.NodeDataLoader(
-        train_g,
-        train_nid,
-        sampler,
-        use_ddp=n_gpus > 1,
-        device=dev_id if args.num_workers == 0 else None,
-        batch_size=args.batch_size,
-        shuffle=True,
-        drop_last=False,
-        num_workers=args.num_workers)
+
+    nccl_comm = None
+    if nccl_id:
+        nccl_comm = nccl.Communicator(n_gpus, proc_id, nccl_id)
+
+    if not args.data_cpu:
```

Comment:
Do you put the node feature into GPUTensor in this example?

Reply:
Yes, the MultiGPUNodeDataLoader stores all node features split on the GPUs behind the scenes.